### ISSUE TITLE
Explain biometic unlock in great(er) detail in the wikis credential section

### ISSUE THREAD

**Issue by Xnyle** [Kunzisoft/KeePassDX#807]
Basically what this all is about it that the biometric unlock feature is a bit voodoo to me.

I tried to look up what's actually happening, found something here (maybe that could be taken as a basis)
https://support.1password.com/android-biometric-unlock-security/
But that leaves way to many new questions:

KeePassDX creates a second encrypted master key for the database, right? but that's not stored inside the db cause the file is not changed?
So where is that key stored then?

Then that key is encrypted with some android biometric unlock voodoo and here i get completely lost.

What is actually stored where and how is it verified? The 1password doc i mentioned above just states "the Authenticated Key is never stored on your device" Well how is that even possible? You can't possible create that out of random bits of my thumb, can you?

I have to know how things work in order to really trust (or not trust) them. And I think it should be clearly documented how biometic unlock actually works.

**Comment by J-Jamet** [comment]
This is obviously not voodoo.
First of all, you have access to the source code and you can check whether the implementation is right for you or not! (What 1Password does not offer.)
https://github.com/Kunzisoft/KeePassDX/tree/master/app/src/main/java/com/kunzisoft/keepass/biometric

Basically, your link explains rather well how fingerprint unlocking works.
KeePassDX uses the phone [keystore](https://developer.android.com/training/articles/keystore) to generate a random key, this random key is well protected and only validated by the [biometric API](https://developer.android.com/training/sign-in/biometric-auth) if the recognition is OK, then uses this authentication (with a [CryptoObject](https://developer.android.com/reference/android/hardware/biometrics/BiometricPrompt.CryptoObject)) to encrypt or decrypt the master key. This encrypted content is stored in a database internal to the application (different from your database kdbx file)

_"the Authenticated Key is never stored on your device"_ means that the authentication key generated through the keystore and biometric recognition is used on the fly to encrypt and decrypt the content and is not stored.

> I have to know how things work in order to really trust (or not trust) them. And I think it should be clearly documented how biometic unlock actually works.

I understand and that's why it's open source. And for the most paranoid, I am currently thinking about going even further by offering future options to remove encrypted content after a predetermined period of time or after a recognition failure.
https://github.com/Kunzisoft/KeePassDX/issues/566
https://github.com/Kunzisoft/KeePassDX/issues/548

**Comment by Xnyle** [comment]
Thanks for the quick response, though I think there might be a misunderstanding. 
Yes I can look at the source and yes, closed source is bad.

But my question was more about the following:
"this random key is well protected and only validated by the biometric API if the recognition is OK"

This is out of the scope of your app, still It should be explained how this is actually well protected. And this is also the Voodoo I was referring to.

It's probably explained somewhere deep in some Android doc and then even deeper in some hardware implementation documentation. I'm just saying it would be very helpful If that could be put in the Apps documentation in some more accessible format.

If I understand it correctly:

You store a encrypted copy of the files master key in the app db.
This key is encrypted with a random key.
You trust the OS that the random key is well protected.
If it's not, that's a free master key stored on the device, correct?

Is it possible to add an option to not store the random key on storage at all but only in memory? That's a bit like keeypass2android does it with the "short" unlock key. As long as the app is running you may enter a short key (or in this case use your fingerprint), but only once. If you get it wrong the in memory key is deleted.
This is related to #566 i guess but not exactly the same.

**Comment by J-Jamet** [comment]
> This is out of the scope of your app, still It should be explained how this is actually well protected. And this is also the Voodoo I was referring to.

As you say, it is outside the scope of the app. You can therefore consult the official documentation on how these APIs work.
https://source.android.com/security/biometric
Sorry but I'm not going to copy the official documentation just for your convenience.

> You store a encrypted copy of the files master key in the app db.
This key is encrypted with a random key.
You trust the OS that the random key is well protected.
If it's not, that's a free master key stored on the device, correct?

No, why are you making a shortcut by assuming that if you don't trust your phone's keystore, the key is accessible.
In this case, it's just that you don't trust the manufacturer of your device, which I can understand, but you will have much more serious issues.
And I can assure you that this will be the least of your worries. If you have no faith in your device's OS code implementation, every action you take is dangerous. And it is even a security flaw to type a password in any application normally. So you shouldn't even have used your device.
I refer you once again to the official documentation of each element to find out how they are implemented in your OS.
Keystores have been used in java for a long time and are part of the language, I don't see what the condition of your "it's not" is.

> Is it possible to add an option to not store the random key on storage at all but only in memory? 

This is exactly the subject that is discussed in #566. Please read in full.

>  That's a bit like keeypass2android does it with the "short" unlock key.  As long as the app is running you may enter a short key (or in this case use your fingerprint), but only once. If you get it wrong the in memory key is deleted.

KeePass2Android leaves the database completely open in memory for this technique, is that really what you want? https://github.com/Kunzisoft/KeePassDX/issues/102#issuecomment-636722547
Leaving a database completely open with key encryption in a service is, in my opinion, a lot less secure than encrypting a key in the Keystore and if the database is completely closed, but everyone has their own opinion, obviously.

After that it is an optional feature, nothing requires you to use it. And you can of course delete the keys manually using the crossed out fingerprint icon.

**Comment by Xnyle** [comment]
https://source.android.com/security/biometric describes the high level point of view, not what's really going on or at least I can't read enough between the lines / have to little basic knowledge.

But I agree I you propose anyone to just look harder.  It just would be nice to have a more user friendly version in the doc.

Also an open DB in memory is also a bad idea, I agree.

But furthermore I think it's also a bad idea to store a secondary key on the devices storage, it should only ever exist in memory.

The chain fingerprint -> auth key -> secondary master key should not be persistent. 
Fingerprint and auth key is part of the OS infrastructure. Little to tweak here.
But you could add an option to not store the master key in storage, it would have to be regenerated every time after the app restarts by entering the original master keys password. It's then still encrypted via the keystoore/biometric "voodoo" but not persistent.

If someone gets physical access to your device and a master key is stored, the assumption is, that he will never get access to the key because it's encrypted. What if years later a new vulnerability in the HW implementation is found, easy way to brute force fingerprints, or whatever, The key is still stored and you have all the time in the world to try over and over again.
Now you could argue, another key is stored anyways, the password protected original one. But that key can never be a affected by a HW flaw.

I'm not sure if #102 covers not storing the master key (outside of memory) at all. The three tickets linked at the end only refer to removing an already stored key from the keystore? What if you soft brick your device, or it gets stolen. You can't delete the master key in that case. But if it only exists in memory (encrypted ofc) it would be a lot harder to access it.

**Comment by J-Jamet** [comment]
> But I agree if you propose anyone to just look harder. It just would be nice to have a more user friendly version in the doc.

Of course, I understand. For the moment I am trying to describe the technical elements in the issues as well as possible. If it is useful, it is quite possible to update the wiki.

> The chain fingerprint -> auth key -> secondary master key should not be persistent.
Fingerprint and auth key is part of the OS infrastructure. Little to tweak here.
But you could add an option to not store the master key in storage, it would have to be regenerated every time after the app restarts by entering the original master keys password.

Yes we think about it in #102, the main problem is that the notion of "app restart / close" is not obvious and the only solution to have non-persistent data is to use a foreground service which requires notification (and can eat battery). But as I am just testing this part, I will probably use this option by storing encrypted content only in this service if the setting is enable.
I just updated https://github.com/Kunzisoft/KeePassDX/issues/566.

> I'm not sure if #102 covers not storing the master key (outside of memory) at all. The three tickets linked at the end only refer to removing an already stored key from the keystore? What if you soft brick your device, or it gets stolen. You can't delete the master key in that case. But if it only exists in memory (encrypted ofc) it would be a lot harder to access it.

Either if the service is used to store encrypted data (previously in the app database), the Keystore key alone is useless because it is generated randomly when the biometric recognition is used for the first time.

**Comment by J-Jamet** [comment]
It is possible to manage the encrypted keys in a temporary memory since version 2.9.4

**Comment by J-Jamet** [comment]
The wiki page has been updated with the new features : https://github.com/Kunzisoft/KeePassDX/wiki/Advanced-Unlocking

For technical content, it is mainly the same as:
https://github.com/mozilla-lockwise/lockwise-android/blob/master/docs/architecture/sec-apis.md#using-keyguardmanager
but in addition I added a temporary memory managed by a service (for users like you who have this need) and only the Biometric API is used, not the fingerprint API.
